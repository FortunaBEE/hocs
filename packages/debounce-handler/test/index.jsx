import React from 'react';
import { mount } from 'enzyme';

import debounceHandler from '../src/';

const Target = () => null;

describe('debounceHandler', () => {
  it('should pass handler arguments through', (done) => {
    const EnhancedTarget = debounceHandler('testHandler')(Target);
    const mockTestHandler = jest.fn();
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler}/>
    );
    const testHandler = wrapper.find(Target).prop('testHandler');

    testHandler('a', 'b', 'c');
    setTimeout(() => {
      expect(mockTestHandler.mock.calls).toMatchSnapshot();
      done();
    });
  });

  it('should call `e.persist()` if it has been passed', (done) => {
    const EnhancedTarget = debounceHandler('testHandler')(Target);
    const mockTestHandler = jest.fn();
    const mockPersist = jest.fn();
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler}/>
    );
    const testHandler = wrapper.find(Target).prop('testHandler');

    testHandler({ persist: mockPersist });
    setTimeout(() => {
      expect(mockTestHandler.mock.calls).toMatchSnapshot();
      expect(mockPersist.mock.calls).toMatchSnapshot();
      done();
    }, 0);
  });

  it('should debounce handler with `delay` option', (done) => {
    const EnhancedTarget = debounceHandler('testHandler', 50)(Target);
    const mockTestHandler = jest.fn();
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler}/>
    );
    const testHandler = wrapper.find(Target).prop('testHandler');

    testHandler('a');

    setTimeout(() => {
      testHandler('b');

      setTimeout(() => {
        testHandler('c');

        setTimeout(() => {
          testHandler('d');

          setTimeout(() => {
            testHandler('e');

            setTimeout(() => {
              testHandler('f');

              setTimeout(() => {
                expect(mockTestHandler.mock.calls).toMatchSnapshot();
                done();
              }, 50);
            }, 50);
          }, 40);
        }, 30);
      }, 30);
    }, 30);
  });

  it('should debounce handler with `leadingCall` option', (done) => {
    const EnhancedTarget = debounceHandler('testHandler', 50, true)(Target);
    const mockTestHandler = jest.fn();
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler}/>
    );
    const testHandler = wrapper.find(Target).prop('testHandler');

    testHandler('a');

    setTimeout(() => {
      testHandler('b');

      setTimeout(() => {
        testHandler('c');

        setTimeout(() => {
          testHandler('d');

          setTimeout(() => {
            testHandler('e');

            setTimeout(() => {
              testHandler('f');

              setTimeout(() => {
                expect(mockTestHandler.mock.calls).toMatchSnapshot();
                done();
              }, 50);
            }, 50);
          }, 20);
        }, 20);
      }, 20);
    }, 20);
  });

  describe('display name', () => {
    const origNodeEnv = process.env.NODE_ENV;

    afterEach(() => {
      process.env.NODE_ENV = origNodeEnv;
    });

    it('should wrap display name in non-production env', () => {
      process.env.NODE_ENV = 'test';

      const EnhancedTarget = debounceHandler('test')(Target);
      const wrapper = mount(
        <EnhancedTarget/>
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should not wrap display name in production env', () => {
      process.env.NODE_ENV = 'production';

      const EnhancedTarget = debounceHandler('test')(Target);
      const wrapper = mount(
        <EnhancedTarget/>
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
