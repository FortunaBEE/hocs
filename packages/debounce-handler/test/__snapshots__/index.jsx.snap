// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debounceHandler display name should not wrap display name in production env 1`] = `
<DebounceHandler>
  <Target
    test={[Function]}
  />
</DebounceHandler>
`;

exports[`debounceHandler display name should wrap display name in non-production env 1`] = `
<debounceHandler(Target)>
  <Target
    test={[Function]}
  />
</debounceHandler(Target)>
`;

exports[`debounceHandler should call \`e.persist()\` if it has been passed 1`] = `
Array [
  Array [
    Object {
      "persist": [Function],
    },
  ],
]
`;

exports[`debounceHandler should call \`e.persist()\` if it has been passed 2`] = `
Array [
  Array [],
]
`;

exports[`debounceHandler should debounce handler with \`delay\` option 1`] = `
Array [
  Array [
    "e",
  ],
  Array [
    "f",
  ],
]
`;

exports[`debounceHandler should debounce handler with \`leadingCall\` option 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "e",
  ],
  Array [
    "f",
  ],
]
`;

exports[`debounceHandler should pass handler arguments through 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
  ],
]
`;
